src/redisearch_rs/trie_rs/src/trie.rs

```diff
@@ -0,0 +1,917 @@
+                self.children = child.children;
+                self.data = child.data;
+            } else {
+                // If the child's label can fit self's label,
+                // we swap self and child, and prepend the child's label to self's label.
+                let old_self = std::mem::replace(self, child);
+                self.label.prepend_with_slice(&old_self.label);
+            }
+        }
+    }
+
+    /// Split the label of the current node at the given index.
+    /// Then assign the first chunk of the label to the current node,
+    /// while the suffix is assigned to a newly-created child node.
+    ///
+    /// If the current node had any data attached to it, we re-assign
+    /// data to the new child node.
+    ///
+    /// ```text
+    /// // Splitting at index 4
+    /// biker (A)  ->   bike (-)
+    ///                /
+    ///               r (A)
+    /// ```
+    ///
+    /// If the current node has any children, they become children
+    /// of the new child node.
+    ///
+    /// ```text
+    /// // Splitting at index 1
+    /// bi (A)   ->    b (-)
+    ///   \             \
+    ///    ke (B)        i (A)
+    ///                   \
+    ///                    ke (B)
+    /// ```
+    fn split(&mut self, index: usize) {
+        let suffix = LowMemoryThinVec::from_slice(&self.label.as_slice()[index..]);
+        let suffix_first_byte = suffix[0];
+
+        let new_node = Node {
+            children: std::mem::take(&mut self.children),
+            data: self.data.take(),
+            label: suffix,
+        };
+
+        let parent_children = {
+            let child_ref = ChildRef {
+                first_byte: suffix_first_byte,
+                node: new_node,
+            };
+            ChildRefs(low_memory_thin_vec![child_ref])
+        };
+
+        self.label.truncate(index);
+        self.children = parent_children;
+    }
+
+    /// Get a reference to the value associated with a key.
+    /// Returns `None` if the key is not present.
+    fn find(&self, key: &[c_char]) -> Option<&Data> {
+        let suffix = key.strip_prefix(self.label.as_slice())?;
+        let Some(first_byte) = suffix.first() else {
+            // The suffix is empty, so the key and the label are equal.
+            return self.data.as_ref();
+        };
+        self.children.find(*first_byte)?.find(suffix)
+    }
+
+    fn mem_usage(&self) -> usize {
+        self.label.mem_usage()
+            + self.children.0.mem_usage()
+            + self
+                .children
+                .0
+                .iter()
+                .map(|c| c.node.mem_usage())
+                .sum::<usize>()
+    }
+
+    fn label_to_string_lossy(&self) -> String {
+        let slice = self.label.iter().map(|&c| c as u8).collect::<Vec<_>>();
+        String::from_utf8_lossy(&slice).into_owned()
+    }
+}
+
+impl<Data: fmt::Debug> fmt::Debug for Node<Data> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
```

Comment:
Cool :)
Note for the future: We should probably use this for debugability (today we have an implementation  in our debug command that is using iterator, but I believe that this "raw" scan is even better